### PR TITLE
z3 opam package update

### PR DESCRIPTION
**Pull request by philzook58**
Updates to the opam z3 packages, including an experimental nightly package
Please look these over @c-cube

### CODE REVIEW

File: packages/z3/z3.4.8.6/opam [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+license: "MIT"
+dev-repo: "git+https://github.com/Z3prover/z3.git"
+build: [
+  [ "python2.7" "scripts/mk_make.py" "--ml" "--staticlib" ]
+  [ make "-C" "build" "-j" jobs ]
+]
+install: [
+  [ "sh" "-c" "ocamlfind install z3 build/api/ml/META -nodll build/libz3* build/api/ml/*" ]
+]
+remove: ["ocamlfind" "remove" "z3"]
```
Comment (kit-ty-kate) [inline packages/z3/z3.4.8.6/opam]:
```suggestion
```
The `remove` field is optional since opam 2.0

File: packages/z3/z3.4.8.6/opam [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+depends: [
+  "ocaml"
+  "ocamlfind" {build}
+  "num"
+  "conf-gmp"
+  "conf-python-2-7" {build}
+]
+synopsis: "Z3 solver"
+flags: light-uninstall
```
Comment (kit-ty-kate) [inline packages/z3/z3.4.8.6/opam]:
```suggestion
```
If the remove field is removed you won't need that

File: packages/z3/z3.~nightly/opam [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+license: "MIT"
+dev-repo: "git+https://github.com/Z3prover/z3.git"
+build: [
+  [ "python2.7" "scripts/mk_make.py" "--ml" "--staticlib" ]
+  [ make "-C" "build" "-j" jobs ]
+]
+install: [
+  [ "sh" "-c" "ocamlfind install z3 build/api/ml/META -nodll build/libz3* build/api/ml/*" ]
+]
+remove: ["ocamlfind" "remove" "z3"]
```
Comment (kit-ty-kate) [inline packages/z3/z3.~nightly/opam]:
```suggestion
```

File: packages/z3/z3.~nightly/opam [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+depends: [
+  "ocaml"
+  "ocamlfind" {build}
+  "zarith"
+  "conf-gmp"
+  "conf-python-2-7" {build}
+]
+synopsis: "Z3 solver"
+flags: light-uninstall
```
Comment (kit-ty-kate) [inline packages/z3/z3.~nightly/opam]:
```suggestion
```

**Comment by kit-ty-kate** [comment]
I don't think having `z3.~nightly` is a good idea. I think we even have rules against that.
The exact same thing can be gotten with just `opam pin add --dev z3`. It'll take the current git master listed in `dev-repo`. This would be the more classical way of doing this.

**Comment by philzook58** [comment]
I made the requested changes. The dependencies of the nightly version have changed though. It switch from num to zarith.

**Comment by kit-ty-kate** [comment]
> I made the requested changes. The dependencies of the nightly version have changed though. It switch from num to zarith.

mh, and I see there is no opam file in the repository itself, this is indeed an issue. Usually the problem is fixed by just modifying the opam file in the repository accordingly, this way the changes would be picked up by opam. Is there no way of adding an opam file to the z3 repository?

**Comment by philzook58** [comment]
I can ask. I rather doubt they would want an opam file at the top level of the project. Ocaml is not a  priority of the project. 

Would opam be able to find the opam file deeper in the hierarchy somehow? I think maybe they might accept one in this directory https://github.com/Z3Prover/z3/tree/master/src/api/ml

In any case, while I need the latest z3 fixes, not everyone does. So I think if the new z3.4.8.6 seems acceptable as is, it is still useful. And this particular dependency issue should be easily resolved when the next release of z3 is cut even if I can't get an opam file into z3 itself.

**Comment by c-cube** [comment]
I made the suggestion to have a `z3-nightly` alternative package, for people who need it. That'd be less misleading to casual users.

**Comment by kit-ty-kate** [comment]
Thanks!

**Pull request merged**